// @airtable/blocks/dist/cjs/models/field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.symbol");

require("core-js/modules/es.symbol.description");

require("core-js/modules/es.array.filter");

require("core-js/modules/es.array.iterator");

require("core-js/modules/es.object.to-string");

require("core-js/modules/es.set");

require("core-js/modules/web.dom-collections.iterator");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _field = require("../types/field");

var _private_utils = require("../private_utils");

var _get_sdk = _interopRequireDefault(require("../get_sdk"));

var _abstract_model = _interopRequireDefault(require("./abstract_model"));

var _aggregators = _interopRequireDefault(require("./aggregators"));

/** @module @airtable/blocks/models: Field */

/** */
// This doesn't follow our enum naming conventions because we want the keys
// to mirror the method/getter names on the model class.
var WatchableFieldKeys = Object.freeze({
  name: 'name',
  type: 'type',
  options: 'options',
  isComputed: 'isComputed',
  description: 'description'
});
/**
 * All the watchable keys in a field.
 * - `name`
 * - `type`
 * - `options`
 * - `isComputed`
 * - `description`
 */

/**
 * Model class representing a field in a table.
 *
 * @example
 * ```js
 * import {base} from '@airtable/blocks';
 *
 * const table = base.getTableByName('Table 1');
 * const field = table.getFieldByName('Name');
 * console.log('The type of this field is', field.type);
 * ```
 * @docsPath models/Field
 */
var Field =
/*#__PURE__*/
function (_AbstractModel) {
  (0, _inherits2.default)(Field, _AbstractModel);
  (0, _createClass2.default)(Field, null, [{
    key: "_isWatchableKey",

    /** @internal */

    /** @internal */
    value: function _isWatchableKey(key) {
      return (0, _private_utils.isEnumValue)(WatchableFieldKeys, key);
    }
    /** @internal */

  }]);

  /**
   * @internal
   */
  function Field(baseData, parentTable, fieldId) {
    var _this;

    (0, _classCallCheck2.default)(this, Field);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Field).call(this, baseData, fieldId));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_parentTable", void 0);
    _this._parentTable = parentTable;
    return _this;
  }
  /**
   * @internal
   */


  (0, _createClass2.default)(Field, [{
    key: "isAggregatorAvailable",

    /**
     * Checks if the given aggregator is available for this field.
     *
     * @param aggregator The aggregator object or aggregator key.
     * @example
     * ```js
     * import {aggregators} from '@airtable/blocks/models';
     * const aggregator = aggregators.totalAttachmentSize;
     *
     * // Using an aggregator object
     * console.log(myAttachmentField.isAggregatorAvailable(aggregator));
     * // => true
     *
     * // Using an aggregator key
     * console.log(myTextField.isAggregatorAvailable('totalAttachmentSize'));
     * // => false
     * ```
     */
    value: function isAggregatorAvailable(aggregator) {
      var aggregatorKey = typeof aggregator === 'string' ? aggregator : aggregator.key;
      var airtableInterface = this._parentTable._airtableInterface;
      var availableAggregatorKeys = airtableInterface.aggregators.getAvailableAggregatorKeysForField(this._data);
      return availableAggregatorKeys.some(key => key === aggregatorKey);
    }
    /**
     * Attempt to parse a given string and return a valid cell value for the field's current config.
     * Returns `null` if unable to parse the given string.
     *
     * @param string The string to parse.
     * @example
     * ```js
     * const inputString = '42';
     * const cellValue = myNumberField.convertStringToCellValue(inputString);
     * console.log(cellValue === 42);
     * // => true
     * ```
     */

  }, {
    key: "convertStringToCellValue",
    value: function convertStringToCellValue(string) {
      var airtableInterface = (0, _get_sdk.default)().__airtableInterface;

      var appInterface = (0, _get_sdk.default)().__appInterface;

      var cellValue = airtableInterface.fieldTypeProvider.convertStringToCellValue(appInterface, string, this._data); // Temporarily bail out of validating computed values (since validation will crash)
      // while we work out if we actually have to validate or not. Ideally we just delete all
      // the validation
      // TODO(emma): delete me or tidy me up

      if (this.isComputed) {
        return cellValue;
      } // TODO(emma): do we need to validate here?


      var validationResult = airtableInterface.fieldTypeProvider.validateCellValueForUpdate(appInterface, cellValue, null, this._data);

      if (validationResult.isValid) {
        return cellValue;
      } else {
        return null;
      }
    }
    /**
     * @internal
     */

  }, {
    key: "__getRawType",
    value: function __getRawType() {
      return this._data.type;
    }
    /**
     * @internal
     */

  }, {
    key: "__getRawTypeOptions",
    value: function __getRawTypeOptions() {
      return this._data.typeOptions;
    }
    /**
     * @internal
     */

  }, {
    key: "__triggerOnChangeForDirtyPaths",
    value: function __triggerOnChangeForDirtyPaths(dirtyPaths) {
      if (dirtyPaths.name) {
        this._onChange(WatchableFieldKeys.name);
      }

      if (dirtyPaths.type) {
        this._onChange(WatchableFieldKeys.type); // TODO: it would be better if we only trigger this when
        // we know isComputed changed.


        this._onChange(WatchableFieldKeys.isComputed);
      }

      if (dirtyPaths.typeOptions) {
        this._onChange(WatchableFieldKeys.options);
      }

      if (dirtyPaths.description) {
        this._onChange(WatchableFieldKeys.description);
      }
    }
  }, {
    key: "_dataOrNullIfDeleted",
    get: function get() {
      var _ref;

      var tableData = this._baseData.tablesById[this.parentTable.id];
      return (_ref = tableData === null || tableData === void 0 ? void 0 : tableData.fieldsById[this._id]) !== null && _ref !== void 0 ? _ref : null;
    }
    /**
     * The table that this field belongs to. Should never change because fields aren't moved between tables.
     *
     * @internal (since we may not be able to return parent model instances in the immutable models world)
     * @example
     * ```js
     * const field = myTable.getFieldByName('Name');
     * console.log(field.parentTable.id === myTable.id);
     * // => true
     * ```
     */

  }, {
    key: "parentTable",
    get: function get() {
      return this._parentTable;
    }
    /**
     * The name of the field. Can be watched.
     *
     * @example
     * ```js
     * console.log(myField.name);
     * // => 'Name'
     * ```
     */

  }, {
    key: "name",
    get: function get() {
      return this._data.name;
    }
    /**
     * The type of the field. Can be watched.
     *
     * @example
     * ```js
     * console.log(myField.type);
     * // => 'singleLineText'
     * ```
     */

  }, {
    key: "type",
    get: function get() {
      var airtableInterface = (0, _get_sdk.default)().__airtableInterface;

      var appInterface = (0, _get_sdk.default)().__appInterface;

      var _airtableInterface$fi = airtableInterface.fieldTypeProvider.getConfig(appInterface, this._data, this.parentTable.__getFieldNamesById()),
          type = _airtableInterface$fi.type; // We intend to switch from "lookup" to "multipleLookupValues", but need to support both
      // until the transition is complete. See <https://airtable.quip.com/VxaMAmAfUscs> for more.
      // @ts-ignore


      if (type === 'lookup') {
        return _field.FieldType.MULTIPLE_LOOKUP_VALUES;
      } else {
        return type;
      }
    }
    /**
     * The configuration options of the field. The structure of the field's
     * options depend on the field's type. Can be watched.
     *
     * @see {@link FieldType}
     * @example
     * ```js
     * import {FieldType} from '@airtable/blocks/models';
     *
     * if (myField.type === FieldType.CURRENCY) {
     *     console.log(myField.options.symbol);
     *     // => '$'
     * }
     * ```
     */

  }, {
    key: "options",
    get: function get() {
      var airtableInterface = (0, _get_sdk.default)().__airtableInterface;

      var appInterface = (0, _get_sdk.default)().__appInterface;

      var _airtableInterface$fi2 = airtableInterface.fieldTypeProvider.getConfig(appInterface, this._data, this.parentTable.__getFieldNamesById()),
          options = _airtableInterface$fi2.options; // TODO(emma): can we remove this cloneDeep?


      return options ? (0, _private_utils.cloneDeep)(options) : null;
    }
    /**
     * `true` if this field is computed, `false` otherwise. A field is
     * "computed" if it's value is not set by user input (e.g. autoNumber, formula,
     * etc.). Can be watched
     *
     * @example
     * ```js
     * console.log(mySingleLineTextField.isComputed);
     * // => false
     * console.log(myAutoNumberField.isComputed);
     * // => true
     * ```
     */

  }, {
    key: "isComputed",
    get: function get() {
      var airtableInterface = (0, _get_sdk.default)().__airtableInterface;

      return airtableInterface.fieldTypeProvider.isComputed(this._data);
    }
    /**
     * `true` if this field is its parent table's primary field, `false` otherwise.
     * Should never change because the primary field of a table cannot change.
     */

  }, {
    key: "isPrimaryField",
    get: function get() {
      return this.id === this.parentTable.primaryField.id;
    }
    /**
     * The description of the field, if it has one. Can be watched.
     *
     * @example
     * ```js
     * console.log(myField.description);
     * // => 'This is my field'
     * ```
     */

  }, {
    key: "description",
    get: function get() {
      return this._data.description;
    }
    /**
     * A list of available aggregators given this field's configuration.
     *
     * @example
     * ```js
     * const fieldAggregators = myField.availableAggregators;
     * ```
     */

  }, {
    key: "availableAggregators",
    get: function get() {
      var airtableInterface = this._parentTable._airtableInterface;
      var availableAggregatorKeysSet = new Set(airtableInterface.aggregators.getAvailableAggregatorKeysForField(this._data));
      return (0, _private_utils.values)(_aggregators.default).filter(aggregator => {
        return availableAggregatorKeysSet.has(aggregator.key);
      });
    }
  }]);
  return Field;
}(_abstract_model.default);

(0, _defineProperty2.default)(Field, "_className", 'Field');
var _default = Field;
exports.default = _default;
